### PR TITLE
Cpumanager works after kubelet restart

### PR DESCRIPTION
**Pull request by flyingcougar**
**What this PR does / why we need it**:
Cpumanager after restart looses cpu assignment. Guaranteed pods are being assigned to default cpuset. Implementation has been split as follows:

- [x] File backed `State` implementation #54408
- [x] Enabling of file backed  `State` in policy #54409
- [ ] Reconcile loop can restore state #54410

Part of #52031

### CODE REVIEW

**Comment by flyingcougar** [comment]
Additional note:
During pod deletion, `RemoveContainer` is called, then reconcile loop kicks is, finds the deleted container and because it is no present in state, default cpu set is assigned during container termination process.

In added check if that case is found (pod is running but `DeletionTimestamp` is present) container cpuset update is skipped

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 1]
```diff
@@ -228,6 +228,23 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
 				continue
 			}
 
+			// Check whether container is present in state, there may be 2 reasons why it's not present
+			// either container is Be/Bu or kubelet just restarted
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: Add 3rd reason here (per your note in the PR comments) which is the container might be in the process of being destroyed. 

> Additional note:
> During pod deletion, RemoveContainer is called, then reconcile loop kicks is, finds the deleted container and because it is not present in state, default cpuset is assigned during container termination process.

> In added check if that case is found (pod is running but DeletionTimestamp is present) container cpuset update is skipped

File: pkg/kubelet/cm/cpumanager/policy_static.go [hunk 2]
```diff
@@ -123,6 +123,12 @@ func (p *staticPolicy) AddContainer(s state.State, pod *v1.Pod, container *v1.Co
 	glog.Infof("[cpumanager] static policy: AddContainer (pod: %s, container: %s, container id: %s)", pod.Name, container.Name, containerID)
 	if numCPUs := guaranteedCPUs(pod, container); numCPUs != 0 {
 		// container belongs in an exclusively allocated pool
+
+		if _, ok := s.GetCPUSet(containerID); ok {
+			glog.Infof("[cpumanager] static policy: container: %s, container id: %s already present in state, skipping", container.Name, containerID)
```
[the remark this answers is not in the document]
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: ok

[the remark this answers is not in the document]
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: ok

File: pkg/kubelet/cm/cpumanager/policy_static_test.go [hunk 3]
```diff
@@ -37,6 +37,8 @@ type staticPolicyTest struct {
 	expErr          error
 	expCPUAlloc     bool
 	expCSet         cpuset.CPUSet
+	keepPrevPolicy  bool
```
[the remark this answers is not in the document]
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/policy_static_test.go]: ok

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 1 above)
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: ok

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 4]
```diff
@@ -228,6 +228,19 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
...
+			if _, ok := m.state.GetCPUSet(containerID); !ok {
+				if status.Phase == v1.PodRunning && pod.DeletionTimestamp == nil {
```
[the remark this answers is not in the document]
Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: ok

**Comment by ConnorDoyle** [comment]
@flyingcougar were you able to re-run the node e2e for CPU manager with this PR? Since it doesn't run as part of the gates yet I think we need to verify it manually for now.

`make test-e2e-node TEST_ARGS='--feature-gates=DynamicKubeletConfig=true' FOCUS="CPU Manager" SKIP="" PARALLELISM=1`

**Comment by flyingcougar** [comment]
@ConnorDoyle Re run the tests locally and all works fine.

**Comment by flyingcougar** [comment]
cc @derekwaynecarr  @jeremyeder

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 5]
```diff
@@ -228,6 +228,26 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
...
 
+			// Check whether container is present in state, there may be 3 reasons why it's not present:
+			// - container is in Be/Bu pod
+			// - kubelet just restarted - state is empty
+			// - container has been removed from state by RemoveContainer call (DeletionTimestamp is set)
+			if _, ok := m.state.GetCPUSet(containerID); !ok {
+				if status.Phase == v1.PodRunning && pod.DeletionTimestamp == nil {
+
+					glog.Infof("[cpumanager] reconcileState: container is not present in state - trying to add (pod: %s, container: %s, container id: %s)", pod.Name, container.Name, containerID)
+					err := m.AddContainer(pod, &container, containerID)
```
Comment (balajismaniam) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: @flyingcougar I don't think this is the right approach to handle Kubelet restarts for Gu pods. `AddContainer` may assign different set of CPUs than what was previously assigned via `allocateCPUs`->`takeByTopology`. If the container is running we should inspect it's cpuset and assign the previously assigned CPUs directly using `SetCPUSet` or some other mechanism. 

@ConnorDoyle WDYT?

Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: Agree. The cpuset may be different than it was before the kubelet restart. The "size" of cpuset will remain consistent. With i.e cgroup inspection we can retain the original cpuset.

**Comment by ConnorDoyle** [comment]
> If the container is running and if it is a Gu pod, we should inspect it's cpuset and assign the previously assigned CPUs

Reading the cgroup controller files isn't a great solution to me, since those are "owned" by the CRI implementation. You can imagine a container runtime based on something other than cgroup.

I think a better approach would be to provide a persistent (local disk backed) implementation of the `state.State` interface. This preserves the ownership semantics across Kubelet restarts better in my opinion, which is that assignments flow from policy to state to CRI. Using cgroupfs as an alternate source of truth seems kind of complicated (how do we reconstruct the state's default cpuset, for example?)

That still leaves the question open about whether we need this PR.

**Comment by balajismaniam** [comment]
> I think a better approach would be to provide a persistent (local disk backed) implementation of the state.State interface. This preserves the ownership semantics across Kubelet restarts better in my opinion, which is that assignments flow from policy to state to CRI. 

Local-disk backed `state.State` sgtm.

**Comment by flyingcougar** [comment]
After some testing `AddContainer` in reconcile loop may be needed after all. In case we manage to pickup checkpoint file and load it's content to `state` all is fine, but what when the file is corrupted or missing ?

In that case i think the appropriate action would be to initialize empty state and figure out in reconcile loop what pods are running and add them to `Policy` & `state` 

@balajismaniam @ConnorDoyle  ?

**Comment by derekwaynecarr** [comment]
apologize for delay getting to this pr, will take a look.

File: pkg/kubelet/cm/cpumanager/state/state_file.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	DefaultCpuSet string            `json:"defaultCpuSet"`
+	Entries       map[string]string `json:"reservedList,omitempty"`
+}
+
+type stateFile struct {
+	sync.RWMutex
+	stateFilePath string
+}
+
+func NewStateFileBackend() *stateFile {
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/state/state_file.go]: can we avoid this and reconcile from the cgroupfs itself?

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
i would like to avoid state files.

can we reconcile from the cgroupfs?

we do this today in other places for handling pod cgroups.

**Comment by ConnorDoyle** [comment]
@derekwaynecarr see this comment from above: https://github.com/kubernetes/kubernetes/pull/52709#issuecomment-333578587

IMO it's ok to make certain assumptions about pod cgroupfs since it is owned by the Kubelet, but the same does not apply to container cgroups.

**Comment by flyingcougar** [comment]
@derekwaynecarr @ConnorDoyle what about `ContainerInspect` call, add it to CRI and use for CPU Manager bootstrap ?
Once 'activePods' function (in `reconcilState`) starts to return some pods:
* if pod/container is in state:
  * if cpuset in state match cpuset from `ContainerInspect` then ok
  * if cpuset is state does not match `ContainerInspect` use state and update
* if pod/container is not in the state (and not being deleted)
 * Try to add thru `AddContainer` policy call

**Comment by dchen1107** [comment]
Can we recover the state from cgroup itself?

**Comment by derekwaynecarr** [comment]
per discussion in wg-resource-mgmt call, we decided to checkpoint akin to device plugin for the time being.

**Comment by flyingcougar** [comment]
@derekwaynecarr - done

File: pkg/kubelet/cm/cpumanager/state/state_mem.go [hunk 7]
```diff
@@ -23,20 +23,39 @@ import (
...
 	defaultCPUSet cpuset.CPUSet
+	stateFile     *stateFile
 }
 
 var _ State = &stateMemory{}
 
 // NewMemoryState creates new State for keeping track of cpu/pod assignment
 func NewMemoryState() State {
-	glog.Infof("[cpumanager] initializing new in-memory state store")
+	glog.Infof("[cpumanager] state memory: initializing new state store")
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_mem.go]: Let's leave the memory-backed state implementation as-is and just implement a new file-backed state.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 8]
```diff
@@ -228,6 +228,26 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
 				continue
 			}
 
+			// Check whether container is present in state, there may be 3 reasons why it's not present:
+			// - container is in Be/Bu pod
+			// - kubelet just restarted - state has not been restored properly and is empty
+			// - container has been removed from state by RemoveContainer call (DeletionTimestamp is set)
+			if _, ok := m.state.GetCPUSet(containerID); !ok {
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: Can we move this new block out to a separate PR and just make this PR about the persistent state?

File: pkg/kubelet/cm/cpumanager/state/state_mem.go [hunk 9]
```diff
@@ -23,20 +23,39 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
 )
 
+type ContainerCpuAssignment map[string]cpuset.CPUSet
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/state/state_mem.go]: s/ContainerCpuAssignment/ContainerCPUAssignments

and move the type definition from state_mem.go to state.go

**Review by ConnorDoyle — COMMENTED** [review]
For sanity's sake I'd love to see this as 3 separate PRs:

1. Introduce the file backed implementation of state.State and tests (leaving the memory state alone, as noted in the comments below.
1. Use the file-backed state by default in the cpu manager, add persistence-related tests for the static policy.
1. Update the reconcile behavior.

File: pkg/kubelet/cm/cpumanager/state/state_mem.go (hunk 7 above)
Comment (balajismaniam) [reply pkg/kubelet/cm/cpumanager/state/state_mem.go]: +1

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 8 above)
Comment (balajismaniam) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: I would also suggest that we find a way to not do `AddContainer` in reconcile loop. May be it is enough if we do a reconcile on kubelet start/re-start if there is a cpu manager checkpoint file. If the checkpoint file is missing or if there are dangling Gu pods, only then we use `AddContainer`. The other option would be more drastic in that if the checkpoint file is missing, we drain the node? WDYT?

Comment (flyingcougar) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: @ConnorDoyle - ok
@balajismaniam 
> if we do a reconcile on kubelet start/re-start

I was thinking the same but then how many time should we run it ? `activePods` function after restart may not report anything for some time, and them might start reporting only some pods (already found by PLEG)

>If the checkpoint file is missing or if there are dangling Gu pods, only then we use AddContainer
CPUManager does not know (not should) know what pods/containers Policy is handling. It's Policy implementation that decides how and whom cpus should assign (Gu in case static Policy)
